### ISSUE TITLE
[DEP_WEBPACK_WATCH_WITHOUT_CALLBACK] error when using `--watch` flag

### ISSUE THREAD

**Issue by sanex3339** [webpack/webpack-cli#1918]
**Describe the bug**

PR with the configuration
https://github.com/javascript-obfuscator/javascript-obfuscator/pull/776

After upgrade from webpack@4 to webpack@5 and webpack-cli@3 to webpack-cli@4.0.0 I got the following error:
```
[DEP_WEBPACK_WATCH_WITHOUT_CALLBACK] DeprecationWarning: A 'callback' argument need to be provided to the 'webpack(options, callback)' function when the 'watch' option is set. There is no way to handle the 'watch' option without a callback.
```

This error happens only when using `--watch` flag:
```
webpack --config ./webpack/webpack.node.config.js --mode development --watch
```

Full error trace:
```
(node:62890) [DEP_WEBPACK_WATCH_WITHOUT_CALLBACK] DeprecationWarning: A 'callback' argument need to be provided to the 'webpack(options, callback)' function when the 'watch' option is set. There is no way to handle the 'watch' option without a callback.
    at webpack (/Users/sanex/WebstormProjects/javascript-obfuscator/node_modules/webpack/lib/webpack.js:145:5)
    at f (/Users/sanex/WebstormProjects/javascript-obfuscator/node_modules/webpack/lib/index.js:35:15)
    at Compiler.createCompiler (/Users/sanex/WebstormProjects/javascript-obfuscator/node_modules/webpack-cli/lib/utils/Compiler.js:129:35)
    at WebpackCLI.run (/Users/sanex/WebstormProjects/javascript-obfuscator/node_modules/webpack-cli/lib/webpack-cli.js:220:32)
    at async runCLI (/Users/sanex/WebstormProjects/javascript-obfuscator/node_modules/webpack-cli/lib/bootstrap.js:74:24)
```

**Expected behavior**

No error.

**Comment by evilebottnawi** [comment]
@webpack/cli-team high priority

**Comment by photonstorm** [comment]
I get the same error, but I think it's worth mentioning that the watch mode does actually work. Yes, it throws that warning at the start, but it still bundles and still watches for changes correctly.

**Comment by raduchiriac** [comment]
Indeed, webpack v5 does not watch anymore if I remove `--watch` to go with `webpack --config webpack.config.js`
I use this:
```js
{
  mode: "development",
  watch: true,
  devtool: "inline-cheap-source-map",
  watchOptions: {
    ignored: ["node_modules/**"],
  },
  entry: {
    main: ["core-js/stable", "./src/js/index.js"],
    theme: "./src/scss/th.scss",
  },
}
```

With my config above the only way to go, is by adding `--watch` which I prefer not to

**Comment by snitin315** [comment]
I will fix it 👍

**Comment by evilebottnawi** [comment]
We have two problems here:
- watch without callback
- we don't respect the `watch` option from config

**Comment by snitin315** [comment]
@evilebottnawi fixed with #1931 ?

**Comment by evilebottnawi** [comment]
@snitin315 No, will be fixed in the near future

**Comment by evilebottnawi** [comment]
Some problems on webpack side with new API, we are working on it, so I will do release with other fixes, but it will be fixed tomorrow with patch release

**Comment by gimmi** [comment]
Same problem for me when trying to upgrade to `webpack@5.1.3` and `webpack-cli@4.0.0`

**Comment by evilebottnawi** [comment]
@gimmi why need to spam it? I've always tried to figure it out, but I can't, can you explain it to me?

**Comment by evilebottnawi** [comment]
Just note, this is unnecessary using `--watch` and `watch: true` together, just remove `watch: true` from the config

**Comment by jakebailey** [comment]
I hit this with the latest versions of everything and I don't have `watch: true` set, so I don't think this config option being duplicated is the whole story.

**Comment by evilebottnawi** [comment]
@jakebailey Can you provide link on example of problem, removing `watch: true` should help

**Comment by jakebailey** [comment]
I haven't yet pushed the changes to my project to grab 4.1.0 quite yet (holding back dep updates for a release, but will likely do it soon), but the config that it'd be is here: https://github.com/microsoft/pyright/blob/master/packages/vscode-pyright/webpack.config.js

Where the task just runs `webpack --mode development --watch --progress`.

**Comment by evilebottnawi** [comment]
@jakebailey Very strange, should be no problems

**Comment by jasonwilliams** [comment]
I agree with @jakebailey i don’t have watch: true in my config and I still get this, if I can put together a reproducible build I will. But it’s happening here https://github.com/styled-components/vscode-styled-components

**Comment by jasonwilliams** [comment]
I have this with this project https://github.com/styled-components/vscode-styled-components/blob/master/package.json and we don’t have `watch: true` set

**Comment by tkurki** [comment]
I just created https://gist.github.com/tkurki/5627039d52ca0248648ed383294f57f6 and with that I get the DeprecationWarning running `npx --watch`. Install gave me 4.1.0.

**Comment by chadleong** [comment]
I am running webpack 5.1.3 and webpack-cli 4.1.0 with 

`"dev": "webpack --mode development --watch ./src/frontend/src/index.js -o ./src/frontend/static/frontend/"`

Still get the same [DEP_WEBPACK_WATCH_WITHOUT_CALLBACK]  error

**Comment by piecyk** [comment]
Had a quick look and removing `watch: true`  from config will not help as `--watch ` is still propagated to options

https://github.com/webpack/webpack/blob/master/lib/webpack.js#L140 

Looks like we would need to refactor how the webpack-cli handles complier ( have some wip already but would require https://github.com/webpack/webpack/pull/11784 )

**Comment by evilebottnawi** [comment]
@piecyk Can you clarify?

**Comment by piecyk** [comment]
> Can you clarify?

@evilebottnawi on which part ? 

webpack-cli creates compiler via https://github.com/webpack/webpack-cli/blob/master/packages/webpack-cli/lib/webpack-cli.js#L219

Using `--watch` will set {watch: true} on `this.compilerConfiguration` that we pass as `option` to webpack. On webpack side we create compiler without callback that's why it logs a warning at https://github.com/webpack/webpack/blob/master/lib/webpack.js#L140

Looks like there is no option to opt out from warning just to pass a callback when calling `webpack(...)`

**Comment by evilebottnawi** [comment]
> Using --watch will set {watch: true} on this.compilerConfiguration that we pass as option to webpack.

We should fix it

Also we can remove `watch: true` from config when you run `webpack --watch` with warnings message

**Comment by piecyk** [comment]
> We should fix it

Yes i agree. But should the fix be to remove `watch` from options what are passed to webpack? Maybe `options.watch` used in some internal logic. Imho we should fix it without patching options.

> Also we can remove watch: true from config when you run webpack --watch with warnings message

Don't follow this, but having `watch: true` in config and running with `--watch` should not change anything.

**Comment by evilebottnawi** [comment]
> Yes i agree. But should the fix be to remove watch from options what are passed to webpack? Maybe options.watch used in some internal logic. Imho we should fix it without patching options.
> Don't follow this, but having watch: true in config and running with --watch should not change anything.

No, having `watch: true` in config means to start watching at once, but we need more logic here, `compiler.watch()` === `{ watch: true }`

**Comment by piecyk** [comment]
@evilebottnawi I was thinking about something like this https://github.com/webpack/webpack-cli/compare/master...piecyk:fix-watch

Almost all test pass, only still need to check analyze-flag.test.js, and one in cache-flags.test.js.

Also there is error code change, as webpack is wrapping with try catch and not propagating errors https://github.com/webpack/webpack/blob/master/lib/webpack.js#L135 there is no easy way to choose between error code 1 vs 2, decided for simplicity to go with 1

**Comment by evilebottnawi** [comment]
@piecyk bad fix, with breaking changes

**Comment by evilebottnawi** [comment]
But idea is good, I will look at this in near future

**Comment by evilebottnawi** [comment]
@piecyk Yes, can you send a PR (don't use fork) and I will finish your work

**Comment by thisVioletHydra** [comment]
```
√ Webpack
  Compiled successfully in 1.55s

[webpack-cli] Compilationdevelopment starting...
(node:10048) [DEP_WEBPACK_WATCH_WITHOUT_CALLBACK] DeprecationWarning: A 'callback' argument need to be provided to the 'webpack(options, callback)' function when the 'watch' option is set. There is no way to handle the 'watch' option without a callback.
(Use `node --trace-deprecation ...` to show where the warning was created)
```

**Comment by nestorvanz** [comment]
I did update to `"webpack": "^5.3.2"` and the error still appears

```
[DEP_WEBPACK_WATCH_WITHOUT_CALLBACK] DeprecationWarning: A 'callback' argument need to be provided to the 'webpack(options, callback)' function when the 'watch' option is set. There is no way to handle the 'watch' option without a callback.
```

**Comment by jdelStrother** [comment]
I think we're still waiting on a new release ( > 4.0.0 ) for this.  If I test against the master branch (which can be done using `yarn add --dev https://gitpkg.now.sh/webpack/webpack-cli/packages/webpack-cli`) it seems to work fine.

**Comment by evilebottnawi** [comment]
Release will be today

**Comment by evilebottnawi** [comment]
Fixed in `webpack-cli@4.2.0`, if somebody faced with the problem again, please open a new issue with reproducible test repo, because we tests most of cases :smile:

**Comment by xccjk** [comment]
> Fixed in `webpack-cli@4.2.0`, if somebody faced with the problem again, please open a new issue with reproducible test repo, because we tests most of cases 😄

I still have this problem!

"webpack": "^5.10.3",
"webpack-cli": "^4.2.0",
"webpack-dev-server": "^3.11.0"

node: v14.15.0
npm: 6.14.8
mac: 10.14.6

[github](https://github.com/xccjk/webpack-demo1)

**Comment by pedro-w** [comment]
@xccjk  I did a fresh checkout of your https://github.com/xccjk/webpack-demo1 and didn't see this problem.

I did see a "DEP_WEBPACK_COMPILATION_ASSETS" but this comes from `html-webpack-plugin`

* node v14.15.1
* npm 6.14.8
* Windows 10 x64

**Comment by xccjk** [comment]
> @xccjk I did a fresh checkout of your https://github.com/xccjk/webpack-demo1 and didn't see this problem.
> 
> I did see a "DEP_WEBPACK_COMPILATION_ASSETS" but this comes from `html-webpack-plugin`
> 
> * node v14.15.1
> * npm 6.14.8
> * Windows 10 x64

Thank you very much for your reply, but after removing `html-webpack-plugin`, I still have the same problem. I updated the github address after removing redundant packages

- node v14.15.0
- npm 6.14.8
- mac 10.14.6

**Comment by lebadapetru** [comment]
`"webpack": "^5.12",` and this still occurs

**Comment by alexander-akait** [comment]
You need update `webpack-cli` to the latest stable version

**Comment by lebadapetru** [comment]
i'm not actually using pure webpack but webpack encore which uses webpack-cli 4, i see that 4.5 is the latest version but i guess i'll have to wait for webpack encore to upgrade it

**Comment by chadleong** [comment]
Running latest webpack 5.22 and webpack-cli 4.5 - No errors !

**Comment by PeterBowater** [comment]
Once I had updated webpack-cli it was all good :)

**Comment by xeroxstar** [comment]
"webpack": "^5.61.0" getting same warning
